File: 2024/05_Print Queue/Day05.js
// Day 5: Print Queue
// https://adventofcode.com/2024/day/5

const fs = require('fs');
var readInput = function (filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8');
};

function toNumArray(input, separator) {
  return input.split('\n').map((line) => line.split(separator).map(Number));
}

function prepareData(input) {
  const [firstSection, secondSection] = input.trim().split('\n\n');
  const rules = toNumArray(firstSection, '|');
  const updates = toNumArray(secondSection, ',');
  return [rules, updates];
}

function isValidOrder(rules, list) {
  const map = list.reduce((map, num, index) => map.set(num, index), new Map());
  for (const [x, y] of rules) {
    const xIndex = map.get(x) ?? -1;
    const yIndex = map.get(y) ?? -1;

    if (xIndex !== -1 && yIndex !== -1 && xIndex > yIndex) {
      return false;
    }
  }
  return true;
}

function findMiddleSum(rules, updates) {
  let middleSum = 0;
  for (const list of updates) {
    if (isValidOrder(rules, list)) {
      const midIndex = list.length >> 1;
      middleSum += list[midIndex];
    }
  }
  return middleSum;
}

function reorderAndSum(rules, updates) {
  let middleSum = 0;
  for (const list of updates) {
    if (!isValidOrder(rules, list)) {
      const sortedList = reorderUnsortedList(list);
      const midIndex = sortedList.length >> 1;
      middleSum += sortedList[midIndex];
    }
  }
  return middleSum;

  function reorderUnsortedList(list) {
    const map = list.reduce((map, num, index) => map.set(num, index), new Map());
    const graph = new Map();
    const inDegree = new Map();

    list.forEach((page) => {
      graph.set(page, []);
      inDegree.set(page, 0);
    });

    for (const [x, y] of rules) {
      if (map.has(x) && map.has(y)) {
        graph.get(x).push(y);
        inDegree.set(y, inDegree.get(y) + 1);
      }
    }

    const queue = [];
    for (const [node, degree] of inDegree) {
      if (degree === 0) queue.push(node);
    }

    const sortedList = [];
    while (queue.length > 0) {
      const node = queue.shift();
      sortedList.push(node);
      for (const neighbor of graph.get(node)) {
        inDegree.set(neighbor, inDegree.get(neighbor) - 1);
        if (inDegree.get(neighbor) === 0) queue.push(neighbor);
      }
    }
    return sortedList;
  }
}

function part1(input) {
  const [rules, updates] = prepareData(input);
  return findMiddleSum(rules, updates);
}

function part2(input) {
  const [rules, updates] = prepareData(input);
  return reorderAndSum(rules, updates);
}

console.log(part1(readInput('Day05'))); // 143
console.log(part1(readInput('Day05_test'))); // 6612

console.log(part2(readInput('Day05'))); // 123
console.log(part2(readInput('Day05_test'))); // 4944

File: 2024/05_Print_Queue/Day05.kt
// Day 5: Print Queue
// https://adventofcode.com/2024/day/5

import kotlin.io.path.Path
import kotlin.io.path.readText
fun readInput(name: String) = Path("$name.txt").readText().trim()

fun main() {
    data class ParsedData(
        val rules: List<Pair<Int, Int>>,
        val updates: List<List<Int>>
    )

    fun prepareInput(data: String): ParsedData {
        val sections = data.split("\n\n")

        val rules = sections[0]
            .lineSequence()
            .map { line ->
                val parts = line.split("|")
                parts[0].toInt() to parts[1].toInt()
            }.toList()

        val updates = sections[1]
            .lineSequence()
            .map { line ->
                line.split(",").map { it.toInt() }
            }.toList()

        return ParsedData(rules, updates)
    }

    fun isValidOrder(rules: List<Pair<Int, Int>>, map: Map<Int, Int>): Boolean {
        return rules.all { (x, y) ->
            val xIndex = map[x] ?: -1
            val yIndex = map[y] ?: -1
            xIndex == -1 || yIndex == -1 || xIndex <= yIndex
        }
    }

    fun reorderUnsortedList(rules: List<Pair<Int, Int>>, list: List<Int>): List<Int> {
        val graph = list.associateWith { mutableListOf<Int>() }
        val inDegree = list.associateWith { 0 }.toMutableMap()

        for ((x, y) in rules) {
            if (x in graph && y in graph) {
                graph[x]?.add(y)
                inDegree[y] = inDegree.getValue(y) + 1
            }
        }

        val queue = ArrayDeque(inDegree.filterValues { it == 0 }.keys)
        val sortedList = mutableListOf<Int>()

        while (queue.isNotEmpty()) {
            val node = queue.removeFirst()
            sortedList.add(node)
            for (neighbor in graph[node].orEmpty()) {
                inDegree[neighbor] = inDegree.getValue(neighbor) - 1
                if (inDegree[neighbor] == 0) {
                    queue.add(neighbor)
                }
            }
        }
        return sortedList
    }

    fun findMiddleSum(rules: List<Pair<Int, Int>>, updates: List<List<Int>>): Int {
        return updates.sumOf { update ->
            val map = update.withIndex().associate { it.value to it.index }
            if (isValidOrder(rules, map)) update[update.size / 2] else 0
        }
    }

    fun reorderAndSum(rules: List<Pair<Int, Int>>, updates: List<List<Int>>): Int {
        return updates.sumOf { update ->
            val map = update.withIndex().associate { it.value to it.index }
            if (isValidOrder(rules, map)) 0
            else {
                val sortedList = reorderUnsortedList(rules, update)
                sortedList[update.size / 2]
            }
        }
    }

    fun part1(input: String): Int {
        val (rules, updates) = prepareInput(input)
        return findMiddleSum(rules, updates)
    }

    fun part2(input: String): Int {
        val (rules, updates) = prepareInput(input)
        return reorderAndSum(rules, updates)
    }

    var input = readInput("Day05")
    println(part1(input)) // 143
    println(part2(input)) // 123

    input = readInput("Day05_test")
    println(part1(input)) // 6612
    println(part2(input)) // 4944
}
